File: Tree/111_MinimumDepthOfBinaryTree.py
# coding: utf8


"""
    题目链接: https://leetcode.com/problems/minimum-depth-of-binary-tree/description.
    题目描述:

    Given a binary tree, find its minimum depth.

    The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.

"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        # return self.recursive_tree_of_min_depth(root)
        return self.iterative_tree_of_min_depth(root)

    # 递归解法, 基本操作
    def recursive_tree_of_min_depth(self, root):
        if not root:
            return 0
        if not root.left:
            return self.recursive_tree_of_min_depth(root.right) + 1
        if not root.right:
            return self.recursive_tree_of_min_depth(root.left) + 1

        return min(self.recursive_tree_of_min_depth(root.left), self.recursive_tree_of_min_depth(root.right)) + 1

    # 非递归解法
    # 类似于二叉树的层序遍历
    def iterative_tree_of_min_depth(self, root):
        if not root:
            return 0

        queue = []
        front = 0
        rear = 1
        depth = 0
        queue.append(root)

        while front < rear:
            last = rear
            while front < last:
                root = queue[front]
                front += 1
                # 叶子节点
                if not root.left and not root.right:
                    return depth + 1
                if root.left:
                    queue.append(root.left)
                    rear += 1
                if root.right:
                    queue.append(root.right)
                    rear += 1

            depth += 1

        return depth

File: Tree/llb/111. Minimum Depth of Binary Tree.cpp
/*
问题描述：
Given a binary tree, find its minimum depth.

The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
*/

/*思考：
因为如果一个节点如果只有左子树或者右子树，我们不能取它左右子树中小的作为深度，因为那样会是0。
我们只有在叶子节点才能判断深度，当左右子树都为空时才是叶子节点。
*/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    /*思考：因为如果一个节点如果只有左子树或者右子树，我们不能取它左右子树中小的作为深度，因为那样会是0。
      我们只有在叶子节点才能判断深度，当左右子树都为空时才是叶子节点。
    */
    
    //思路一：层序遍历，计算深度，直到出现某个节点的左右孩子都为NULL时，终止循环。
    /*
    int minDepth(TreeNode* root) {
        if (root == NULL) 
            return 0;
        
        queue<TreeNode*> q;
        int depth = 0;
        q.push(root);
        while (!q.empty()) {
            depth++;
            int n = q.size();
            for (int i=0; i<n ;++i) {
                TreeNode* pTemp = q.front();
                q.pop();
                if (pTemp->left) 
                    q.push(pTemp->left);
                if (pTemp->right)
                    q.push(pTemp->right);
                if (!pTemp->left && !pTemp->right) 
                    return depth;
            }
        }
    }
    */
    
    //思路二：利用递归的思想。
    int minDepth(TreeNode* root) {
        if (root == NULL)
            return 0;
        if (root->left == NULL)
            return minDepth(root->right) + 1;
        if (root->right == NULL)
            return minDepth(root->left) + 1;
        return min(minDepth(root->left), minDepth(root->right)) + 1;
    }
    
};
